build/syntax.go

```diff
@@ -305,7 +310,7 @@ func (x *SetExpr) Span() (start, end Position) {
 // A TupleExpr represents a tuple literal: (List)
 type TupleExpr struct {
 	Comments
-	Start Position
+	Start Position // can be empty if the tuple has no brackets, e.g. `a, b = x`
```

Comment:
I don't really like this. When modifying the AST, we might lose the positions.
Is there a simple alternative?
 * Maybe have a boolean to say if there are parens?
 * Maybe don't store the information in the AST? We decide if we have parens in the printer. For example, we can always print parens, except for the case the tuple is a l-value.

Reply:
I can implement it differently, but that's how it's implemented in the Skylark-in-Go parser, I just copied it from there.

I don't think positions will be lost, even if we need to add brackets to a bracketless tuple. There's no contract that nodes should have a non-zero `Start` attribute (not all nodes even have it), the `Span` method correctly returns the positions.

Reply:
My complaint is more about the code in the printer: `if !v.Start.IsValid() {`.
Positions can be lost when a users modify the AST or insert new nodes.

Reply:
I can add a boolean attribute but it won't be compatible with the Skylark-in-Go parser. Should I add it both here and there so that the implementations are still as compatible as possible?

Regarding the second approach, bracketless tuples can occur in some other situations, for instance, `a, b = c, d` (right hand side), or `dictionary[a, b]` (not sure if we should enforce brackets here).

Reply:
I'd suggest that the printer should decide if/when tuples are parenthesized. This improves consistency and I don't think it's very controversial. We can try and see how it goes. We can refine the rules later, e.g.

`a, b = (c, d)`
`dic[(a, b)]`

Reply:
That would require making the printer more complicated (which is ok in general, but this pull request focuses on the parser part and is already quite complicated there). The printer currently just prints expressions recursively and doesn't know whether a value is an lvalue or not. Also `a, b = c, d` looks legit to me, I'd prefer not to enforce brackets here.

I agree that if a new tuple node is added to a tree it should have brackets by default (otherwise it can break the semantics) which is not the case now. But fixing it will break compatibility with the skylark-in-go parser. Perhaps we should address the issue after we've merged the parsers?